backend/amd64/simd_selection.ml

```diff
@@ -0,0 +1,116 @@
+(**************************************************************************)
+(*                                                                        *)
+(*                                 OCaml                                  *)
+(*                                                                        *)
+(*             Xavier Leroy, projet Cristal, INRIA Rocquencourt           *)
+(*                                                                        *)
+(*   Copyright 2000 Institut National de Recherche en Informatique et     *)
+(*     en Automatique.                                                    *)
+(*                                                                        *)
+(*   All rights reserved.  This file is distributed under the terms of    *)
+(*   the GNU Lesser General Public License version 2.1, with the          *)
+(*   special exception on linking described in the file LICENSE.          *)
+(*                                                                        *)
+(**************************************************************************)
+
+open Arch
+open Simd
+
+(* This will need to be expanded with the addition of three and four argument
+   operations in AVX2 and AVX512. *)
+type register_behavior =
+  | R_to_R
+  | R_R_to_fst
+  | R_RM_to_fst
+
+let arg i args = List.nth args i
+
+(* Assumes untagged int *)
+let extract_constant args name =
+  match args with
+  | Cmm.Cconst_int (i, _) :: args -> i, args
+  | _ ->
+    Misc.fatal_errorf "Did not get integer constant as the first argument to %s"
+      name
+
+let check_cmp_imm i =
+  if i < 0 || i >= 8
+  then Misc.fatal_errorf "Invalid immediate for caml_sse_float32x4_cmp: %d" i
+
+let check_shuf_imm i =
+  if i < 0 || i >= 256
+  then Misc.fatal_errorf "Invalid immediate for caml_sse_shuffle: %d" i
+
+let select_operation_sse op args dbg =
+  match op with
+  | "caml_sse_float32x4_cmp" ->
+    let i, args = extract_constant args "caml_sse_float32x4_cmp" in
+    check_cmp_imm i;
+    Some (Cmp_f32 i, args)
+  | "caml_sse_float32x4_add" -> Some (Add_f32, args)
+  | "caml_sse_float32x4_sub" -> Some (Sub_f32, args)
+  | "caml_sse_float32x4_mul" -> Some (Mul_f32, args)
+  | "caml_sse_float32x4_div" -> Some (Div_f32, args)
+  | "caml_sse_float32x4_max" -> Some (Max_f32, args)
+  | "caml_sse_float32x4_min" -> Some (Min_f32, args)
+  | "caml_sse_float32x4_rcp" -> Some (Rcp_f32, [arg 0 args; arg 0 args])
+  | "caml_sse_float32x4_rsqrt" -> Some (Rsqrt_f32, [arg 0 args; arg 0 args])
+  | "caml_sse_float32x4_sqrt" -> Some (Sqrt_f32, [arg 0 args; arg 0 args])
```

Comment:
I am quite possibly missing something:
why is it always getting the argument
at index `0`?

Reply:
The intrinsic is `float32x4 -> float32x4`, but the instruction has the same `R_RM_to_fst` format as the others, so it expects two arguments.

Reply:
Well, shouldn't it have only one argument
and a new `R_to_fst` (or even `R`) format?

Reply:
The instruction actually takes two arguments though; https://www.felixcloutier.com/x86/rcpps

Reply:
I am not necessarily pushing for a change,
but while the `amd64` instruction ultimately
has 2 operands, it does not imply that the
`Mach` instruction should have two `arg`s.
From the linked description, I would expect
one of the operand to be in `arg` and the
other one to be in `res`. But then again,
I am not familiar with these instructions.

Reply:
Ah, true - I actually added a `RM_to_R` behavior in #1665, so this can just use that.